lib/Infer/AliveDriver.cpp

```diff
@@ -814,7 +826,114 @@ IR::Type &souper::AliveDriver::getType(int n) {
   }
   return *TypeCache[n];
 }
+namespace souper {
+void collectPhis(souper::Inst *I, std::map<souper::Block *, std::set<souper::Inst *>> &Phis) {
+  std::vector<Inst *> Stack{I};
+  std::unordered_set<Inst *> Visited;
+  while (!Stack.empty()) {
+    auto Current = Stack.back();
+    Stack.pop_back();
+    if (Current->K == Inst::Phi) {
+      Phis[Current->B].insert(Current);
+    }
+    Visited.insert(Current);
+    for (auto Child : Current->Ops) {
+      if (Visited.find(Child) == Visited.end()) {
+        Stack.push_back(Child);
+      }
+    }
+  }
+}
+
+struct RefinementProblem {
+  souper::Inst *LHS;
+  souper::Inst *RHS;
+  souper::Inst *Pre;
+
+  RefinementProblem ReplacePhi(souper::InstContext &IC, std::map<Block *, size_t> &Change) {
+    std::map<souper::Block *, std::set<souper::Inst *>> Phis;
+    collectPhis(LHS, Phis);
+    collectPhis(Pre, Phis);
+
+    if (Phis.empty()) {
+      return *this; // Base case, no more Phis
+    }
+
+    std::map<Inst *, Inst *> InstCache;
+    for (auto Pair : Phis) {
+      for (auto Phi : Pair.second) {
+        InstCache[Phi] = Phi->Ops[Change[Pair.first]];
+      }
+    }
+    std::map<Block *, Block *> BlockCache;
+    std::map<Inst *, llvm::APInt> ConstMap;
+    RefinementProblem Result;
+    Result.LHS = getInstCopy(LHS, IC, InstCache, BlockCache, &ConstMap, false);
+    Result.RHS = getInstCopy(RHS, IC, InstCache, BlockCache, &ConstMap, false);
+    Result.Pre = getInstCopy(Pre, IC, InstCache, BlockCache, &ConstMap, false);
+
+    // Recursively call ReplacePhi, because Result might have Phi`s
+    return Result.ReplacePhi(IC, Change);
+  }
+  bool operator == (const RefinementProblem &P) const {
+    return LHS == P.LHS && RHS == P.RHS && Pre == P.Pre;
+  }
+  struct Hash
+  {
+    std::size_t operator()(const RefinementProblem& P) const
```

Comment:
" &P" rather than "& P"
try to be consistent with these small things or else they distract from your code

Reply:
also, overloading () deserves a comment. what's going on here?

Reply:
&P is addressed in the other PR.
overloading () for a hash struct is one idiomatic way to use custom types in STL containers like unordered_map or unordered_set.